.changeset/mean-horses-kiss.md

```diff
@@ -0,0 +1,26 @@
+---
+'astro': minor
+---
+
+Adds a new `injectTypes()` utility to the Integration API and refactors how type generation works
+
+Integrations can now use a new `injectTypes()` utility in the `astro:config:done` hook. This utility allows an integration to inject types into a user's project. `filename` must end with `".d.ts"` and `content` must be valid TypeScript (it will be formatted).
+
+Under the hood, it will create a file at `/.astro/integrations/<normalized_integration_name>/<normalized_filename>.d.ts` and create references to it. `injectTypes()` returns a URL to the normalized path.
+
+```js
+const path = injectTypes({
+  filename: "types.d.ts",
+  content: "declare module 'virtual:integration' {}"
+})
+console.log(path) // URL
+```
+
+Codegen has been refactored. As a user, you just need to update `src/env.d.ts`:
+
+```diff
+- /// <reference types="astro/client" />
++ /// <reference path="../.astro/types.d.ts" />
+- /// <reference path="../.astro/env.d.ts" />
+- /// <reference path="../.astro/actions.d.ts" />
+```
```

Comment:
Question: Would this be a breaking change for users who run `astro check`? I think some people run that before `astro build`, and that may break their build pipeline.

Reply:
I think it would not be breaking. To be clear:
- it's fine to keep `astro/client`, it would just be duplicated
- `../.astro/types.d.ts` is not breaking since it already exists for CC
- As for any other core reference, it may break but since those features are marked as experimental I think it's fine
- As for `astro:db`, it may break as well?

@Princesseuh do you know if astro check would complain?

Reply:
`astro check` runs the sync step first, so if the syncing "fix" the types then it'll be fine

Reply:
Outdated declarations are currently not removed

Reply:
Ah, but okay. TypeScript does not report an error for triple slashes references that point to non-existing files, so there shouldn't be an error.

Reply:
Thanks for the explanation. Then it seems fine then. Since the db package is also marked a minor (for v0), it's already a breaking release so changing the types path should be fine.